src/Aspire.Dashboard/Model/ResourceViewModel.cs

```diff
@@ -1,27 +1,34 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Collections.Frozen;
 using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
 using Aspire.Dashboard.Utils;
+using Google.Protobuf.WellKnownTypes;
 
 namespace Aspire.Dashboard.Model;
 
-/// <summary>
-/// Base class for immutable snapshots of resource state at a point in time.
-/// </summary>
-public abstract class ResourceViewModel
+public sealed class ResourceViewModel
 {
     public required string Name { get; init; }
+    public required string ResourceType { get; init; }
     public required string DisplayName { get; init; }
     public required string Uid { get; init; }
     public required string? State { get; init; }
     public required DateTime? CreationTimeStamp { get; init; }
     public required ImmutableArray<EnvironmentVariableViewModel> Environment { get; init; }
-    public required ImmutableArray<EndpointViewModel> Endpoints { get; init; }
-    public required ImmutableArray<ResourceServiceSnapshot> Services { get; init; }
+    public required ImmutableArray<EndpointSnapshot> Endpoints { get; init; }
+    public required ImmutableArray<ResourceServiceViewModel> Services { get; init; }
     public required int? ExpectedEndpointsCount { get; init; }
+    public required FrozenDictionary<string, Value> CustomData { get; init; }
 
-    public abstract string ResourceType { get; }
+    internal bool MatchesFilter(string filter)
+    {
+        // TODO let ResourceType define the additional data values we include in searches
```

Comment:
The idea behind this TODO is that whatever the dashboard connects to can send over the `ResourceDataKey` string values that should participate in user search.